x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/mapper/EvaluatorMapper.java

```diff
@@ -60,15 +71,52 @@ interface ToEvaluator {
      * in that it'll always call {@link Expression#fold}, but that's
      * good enough.
      */
-    default Object fold() {
-        return toJavaObject(toEvaluator(e -> driverContext -> new ExpressionEvaluator() {
+    default Object fold(Source source, FoldContext ctx) {
+        ToEvaluator toEvaluator = new ToEvaluator() {
+            @Override
+            public ExpressionEvaluator.Factory apply(Expression expression) {
+                return driverContext -> new ExpressionEvaluator() {
+                    @Override
+                    public Block eval(Page page) {
+                        return fromArrayRow(driverContext.blockFactory(), expression.fold(ctx))[0];
+                    }
+
+                    @Override
+                    public void close() {}
+                };
+            }
+
+            @Override
+            public FoldContext foldCtx() {
+                return ctx;
+            }
+        };
+
+        CircuitBreaker breaker = ctx.circuitBreakerView(source);
+        BigArrays bigArrays = new BigArrays(null, new CircuitBreakerService() {
+            @Override
+            public CircuitBreaker getBreaker(String name) {
+                if (name.equals(CircuitBreaker.REQUEST) == false) {
+                    throw new UnsupportedOperationException();
+                }
+                return breaker;
+            }
+
             @Override
-            public Block eval(Page page) {
-                return fromArrayRow(driverContext.blockFactory(), e.fold())[0];
+            public AllCircuitBreakerStats stats() {
+                throw new UnsupportedOperationException();
             }
 
             @Override
-            public void close() {}
-        }).get(DriverContext.getLocalDriver()).eval(new Page(1)), 0);
+            public CircuitBreakerStats stats(String name) {
+                throw new UnsupportedOperationException();
+            }
+        }, CircuitBreaker.REQUEST).withCircuitBreaking();
```

Comment:
I think this code block could use a comment to describe why we need to do this.

Reply:
Yeah. It's pretty wild, isn't it? I'll add a paragraph explaining what's up.